[False Positive] Regarding adding const to a function that can't be const

### Summary

If I have this little Rust program (probably it's not minimized, but it's short enough):

```rust
#![feature(const_trait_impl)]
#![feature(nonzero_internals)]
#![allow(internal_features)]

#![warn(clippy::all)]
#![warn(clippy::missing_const_for_fn)]

use std::num::{NonZero, ZeroablePrimitive};

//- - - - - - - -

trait NarrowRem<Out: ZeroablePrimitive> {
    fn narrowing_rem(&self, den: NonZero<Out>) -> Out;
}


impl NarrowRem<u32> for u64 {
    #[inline(always)]
    fn narrowing_rem(&self, den: NonZero<u32>) -> u32 {
        (*self % NonZero::from(den)) as _
    }
}

//- - - - - - - -

#[const_trait]
trait MulModCast where Self: ZeroablePrimitive {
    fn mul_mod(self, b: Self, m: NonZero<Self>) -> Self;
}

impl MulModCast for u32 {
    #[inline]
    fn mul_mod(self, b: Self, m: NonZero<Self>) -> Self {
        (u64::from(self) * u64::from(b)).narrowing_rem(m)
    }
}

//- - - - - - - -

const fn pow_mod(mut a: u32, mut x: u32, nz: NonZero<u32>) -> u32
where u32: MulModCast {
    let mut r: u32 = 1;
    while x != 0 {
        if (x & 1) == 1 {
            r = a.mul_mod(r, nz);
        }
        x >>= 1;
        a = a.mul_mod(a, nz);
    }
    r
}

fn main() {
    assert_eq!(pow_mod(19, 11, NonZero::new(7).unwrap()), 3);
}
```

Clippy gives me this warning:

```rust
  --> ...\main.rs:40:1
   |
40 | / fn pow_mod(mut a: u32, mut x: u32, nz: NonZero<u32>) -> u32
41 | | where u32: MulModCast {
42 | |     let mut r: u32 = 1;
43 | |     while x != 0 {
...  |
50 | |     r
51 | | }
   | |_^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
note: the lint level is defined here
  --> ...\main.rs:6:9
   |
6  | #![warn(clippy::missing_const_for_fn)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: make the function `const`
   |
40 | const fn pow_mod(mut a: u32, mut x: u32, nz: NonZero<u32>) -> u32
   | +++++
```

But if I add const to pow_mod, the rustc compiler (1.88.0-nightly 2025-04-19) gives me:

```rust
error[E0277]: the trait bound `u32: ~const MulModCast` is not satisfied
  --> C:\lavoro\metti_a_posto\progettini_in_corso\eulers\many_solutions\test1\src\main.rs:45:17
   |
45 |             r = a.mul_mod(r, nz);
   |                 ^

error[E0277]: the trait bound `u32: ~const MulModCast` is not satisfied
  --> C:\lavoro\metti_a_posto\progettini_in_corso\eulers\many_solutions\test1\src\main.rs:48:13
   |
48 |         a = a.mul_mod(a, nz);
   |             ^
```

(I have other examples of false positives of clippy::missing_const_for_fn).

### Lint Name

_No response_

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:


### Version

```text

```

### Additional Labels

_No response_

--- a/clippy_utils/src/qualify_min_const_fn.rs
+++ b/clippy_utils/src/qualify_min_const_fn.rs
@@ -18,7 +18,7 @@ use rustc_middle::mir::{
 };
 use rustc_middle::traits::{BuiltinImplSource, ImplSource, ObligationCause};
 use rustc_middle::ty::adjustment::PointerCoercion;
-use rustc_middle::ty::{self, GenericArgKind, TraitRef, Ty, TyCtxt};
+use rustc_middle::ty::{self, GenericArgKind, Instance, TraitRef, Ty, TyCtxt};
 use rustc_span::Span;
 use rustc_span::symbol::sym;
 use rustc_trait_selection::traits::{ObligationCtxt, SelectionContext};
@@ -349,7 +349,15 @@ fn check_terminator<'tcx>(
         }
         | TerminatorKind::TailCall { func, args, fn_span: _ } => {
             let fn_ty = func.ty(body, cx.tcx);
-            if let ty::FnDef(fn_def_id, _) = *fn_ty.kind() {
+            if let ty::FnDef(fn_def_id, fn_substs) = fn_ty.kind() {
+                // FIXME: when analyzing a function with generic parameters, we may not have enough information to
+                // resolve to an instance. However, we could check if a host effect predicate can guarantee that
+                // this can be made a `const` call.
+                let fn_def_id = match Instance::try_resolve(cx.tcx, cx.typing_env(), *fn_def_id, fn_substs) {
+                    Ok(Some(fn_inst)) => fn_inst.def_id(),
+                    Ok(None) => return Err((span, format!("cannot resolve instance for {func:?}").into())),
+                    Err(_) => return Err((span, format!("error during instance resolution of {func:?}").into())),
+                };
                 if !is_stable_const_fn(cx, fn_def_id, msrv) {
                     return Err((
                         span,

--- a/tests/ui/missing_const_for_fn/const_trait.fixed
+++ b/tests/ui/missing_const_for_fn/const_trait.fixed
@@ -1,0 +1,36 @@
+#![feature(const_trait_impl)]
+#![warn(clippy::missing_const_for_fn)]
+
+// Reduced test case from https://github.com/rust-lang/rust-clippy/issues/14658
+
+#[const_trait]
+trait ConstTrait {
+    fn method(self);
+}
+
+impl ConstTrait for u32 {
+    fn method(self) {}
+}
+
+impl const ConstTrait for u64 {
+    fn method(self) {}
+}
+
+fn cannot_be_const() {
+    0u32.method();
+}
+
+//~v missing_const_for_fn
+const fn can_be_const() {
+    0u64.method();
+}
+
+// False negative, see FIXME comment in `clipy_utils::qualify_min_const`
+fn could_be_const_but_does_not_trigger<T>(t: T)
+where
+    T: const ConstTrait,
+{
+    t.method();
+}
+
+fn main() {}

--- a/tests/ui/missing_const_for_fn/const_trait.rs
+++ b/tests/ui/missing_const_for_fn/const_trait.rs
@@ -1,0 +1,36 @@
+#![feature(const_trait_impl)]
+#![warn(clippy::missing_const_for_fn)]
+
+// Reduced test case from https://github.com/rust-lang/rust-clippy/issues/14658
+
+#[const_trait]
+trait ConstTrait {
+    fn method(self);
+}
+
+impl ConstTrait for u32 {
+    fn method(self) {}
+}
+
+impl const ConstTrait for u64 {
+    fn method(self) {}
+}
+
+fn cannot_be_const() {
+    0u32.method();
+}
+
+//~v missing_const_for_fn
+fn can_be_const() {
+    0u64.method();
+}
+
+// False negative, see FIXME comment in `clipy_utils::qualify_min_const`
+fn could_be_const_but_does_not_trigger<T>(t: T)
+where
+    T: const ConstTrait,
+{
+    t.method();
+}
+
+fn main() {}

--- a/tests/ui/missing_const_for_fn/const_trait.stderr
+++ b/tests/ui/missing_const_for_fn/const_trait.stderr
@@ -1,0 +1,17 @@
+error: this could be a `const fn`
+  --> tests/ui/missing_const_for_fn/const_trait.rs:24:1
+   |
+LL | / fn can_be_const() {
+LL | |     0u64.method();
+LL | | }
+   | |_^
+   |
+   = note: `-D clippy::missing-const-for-fn` implied by `-D warnings`
+   = help: to override `-D warnings` add `#[allow(clippy::missing_const_for_fn)]`
+help: make the function `const`
+   |
+LL | const fn can_be_const() {
+   | +++++
+
+error: aborting due to 1 previous error
+

